File: backend/app/models/language.py
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel, Field, validator

class SupportedLanguage(str, Enum):
    """Enumeration of supported language codes"""
    ENGLISH = "en"
    HINDI = "hi"
    TAMIL = "ta"
    TELUGU = "te"
    BENGALI = "bn"
    MARATHI = "mr"

class ScriptType(str, Enum):
    """Enumeration of script types"""
    LATIN = "latin"
    DEVANAGARI = "devanagari"
    TAMIL = "tamil"
    TELUGU = "telugu"
    BENGALI = "bengali"

class TextDirection(str, Enum):
    """Text direction enumeration"""
    LTR = "ltr"
    RTL = "rtl"

class Language(BaseModel):
    """Language model with metadata"""
    code: SupportedLanguage
    name: str
    native_name: str
    direction: TextDirection = TextDirection.LTR
    script_type: ScriptType
    
    class Config:
        use_enum_values = True

# Predefined language configurations
LANGUAGE_CONFIGS = {
    SupportedLanguage.ENGLISH: Language(
        code=SupportedLanguage.ENGLISH,
        name="English",
        native_name="English",
        direction=TextDirection.LTR,
        script_type=ScriptType.LATIN
    ),
    SupportedLanguage.HINDI: Language(
        code=SupportedLanguage.HINDI,
        name="Hindi",
        native_name="हिन्दी",
        direction=TextDirection.LTR,
        script_type=ScriptType.DEVANAGARI
    ),
    SupportedLanguage.TAMIL: Language(
        code=SupportedLanguage.TAMIL,
        name="Tamil",
        native_name="தமிழ்",
        direction=TextDirection.LTR,
        script_type=ScriptType.TAMIL
    ),
    SupportedLanguage.TELUGU: Language(
        code=SupportedLanguage.TELUGU,
        name="Telugu",
        native_name="తెలుగు",
        direction=TextDirection.LTR,
        script_type=ScriptType.TELUGU
    ),
    SupportedLanguage.BENGALI: Language(
        code=SupportedLanguage.BENGALI,
        name="Bengali",
        native_name="বাংলা",
        direction=TextDirection.LTR,
        script_type=ScriptType.BENGALI
    ),
    SupportedLanguage.MARATHI: Language(
        code=SupportedLanguage.MARATHI,
        name="Marathi",
        native_name="मराठी",
        direction=TextDirection.LTR,
        script_type=ScriptType.DEVANAGARI
    )
}

class LanguagePair(BaseModel):
    """Language pair model for translation"""
    source: SupportedLanguage
    target: SupportedLanguage
    
    @validator('target')
    def validate_different_languages(cls, v, values):
        if 'source' in values and v == values['source']:
            raise ValueError('Source and target languages must be different')
        return v
    
    class Config:
        use_enum_values = True

def get_supported_languages() -> List[Language]:
    """Get list of all supported languages"""
    return list(LANGUAGE_CONFIGS.values())

def get_language_by_code(code: str) -> Optional[Language]:
    """Get language configuration by code"""
    try:
        lang_code = SupportedLanguage(code)
        return LANGUAGE_CONFIGS.get(lang_code)
    except ValueError:
        return None

def is_language_pair_supported(source: str, target: str) -> bool:
    """Check if a language pair is supported"""
    try:
        source_lang = SupportedLanguage(source)
        target_lang = SupportedLanguage(target)
        return source_lang != target_lang
    except ValueError:
        return False

File: frontend/src/types/language.ts
// Language interface for supported languages
export interface Language {
  code: string;
  name: string;
  nativeName: string;
  direction: 'ltr' | 'rtl';
  scriptType: 'latin' | 'devanagari' | 'tamil' | 'telugu' | 'bengali';
}

// Supported Indian languages configuration
export const SUPPORTED_LANGUAGES: Language[] = [
  {
    code: 'en',
    name: 'English',
    nativeName: 'English',
    direction: 'ltr',
    scriptType: 'latin'
  },
  {
    code: 'hi',
    name: 'Hindi',
    nativeName: 'हिन्दी',
    direction: 'ltr',
    scriptType: 'devanagari'
  },
  {
    code: 'ta',
    name: 'Tamil',
    nativeName: 'தமிழ்',
    direction: 'ltr',
    scriptType: 'tamil'
  },
  {
    code: 'te',
    name: 'Telugu',
    nativeName: 'తెలుగు',
    direction: 'ltr',
    scriptType: 'telugu'
  },
  {
    code: 'bn',
    name: 'Bengali',
    nativeName: 'বাংলা',
    direction: 'ltr',
    scriptType: 'bengali'
  },
  {
    code: 'mr',
    name: 'Marathi',
    nativeName: 'मराठी',
    direction: 'ltr',
    scriptType: 'devanagari'
  }
];

// Language pair interface
export interface LanguagePair {
  source: string;
  target: string;
}

// Get language by code
export const getLanguageByCode = (code: string): Language | undefined => {
  return SUPPORTED_LANGUAGES.find(lang => lang.code === code);
};

// Check if language pair is supported
export const isLanguagePairSupported = (source: string, target: string): boolean => {
  const sourceExists = SUPPORTED_LANGUAGES.some(lang => lang.code === source);
  const targetExists = SUPPORTED_LANGUAGES.some(lang => lang.code === target);
  return sourceExists && targetExists && source !== target;
};
